### PR TITLE
[TR] PIM-4474 asset completeness

### PR DESCRIPTION
**Pull request by jmleroux**

### CODE REVIEW

**Comment by nidup** [comment]
no specs to update?

File: src/Pim/Bundle/CatalogBundle/Doctrine/ORM/CompletenessGenerator.php [hunk 1]
```diff
@@ -112,18 +113,18 @@ protected function generate(array $criteria = array())
      *
      * @param array $criteria
      */
-    protected function prepareCompletePrices($criteria = array())
+    protected function prepareCompletePrices($criteria = [])
     {
-        $cleanupSql = "DROP TABLE IF EXISTS ".self::COMPLETE_PRICES_TABLE."\n";
+        $cleanupSql  = "DROP TABLE IF EXISTS " . self::COMPLETE_PRICES_TABLE . PHP_EOL;
```
Comment (willy-ahva) [inline src/Pim/Bundle/CatalogBundle/Doctrine/ORM/CompletenessGenerator.php]: simple quotes

File: src/Pim/Bundle/CatalogBundle/Doctrine/ORM/CompletenessGenerator.php [hunk 2]
```diff
@@ -112,18 +113,18 @@ protected function generate(array $criteria = array())
...
         $cleanupStmt = $this->connection->prepare($cleanupSql);
         $cleanupStmt->execute();
 
         $sql = $this->getCompletePricesSQL();
         $sql = $this->applyCriteria($sql, $criteria);
 
-        $sql = "CREATE TEMPORARY TABLE ".
-            self::COMPLETE_PRICES_TABLE.
-            " (locale_id int, channel_id int, value_id int, primary key(locale_id, channel_id, value_id)) ".
+        $sql = "CREATE TEMPORARY TABLE " .
```
Comment (willy-ahva) [inline src/Pim/Bundle/CatalogBundle/Doctrine/ORM/CompletenessGenerator.php]: Same pls simple quotes

**Comment by jmleroux** [comment]
It only add or update protected methods, so i don't think

**Comment by nidup** [comment]
add BC break in changelog then GTM

**Comment by jmleroux** [comment]
@nidup no BC break, it's just a PHPDoc correction

**Comment by grena** [comment]
GTM

File: src/Pim/Bundle/CatalogBundle/Doctrine/ORM/CompletenessGenerator.php [hunk 3]
```diff
@@ -346,10 +345,12 @@ protected function getMainSqlPart()
      */
     protected function getQueryPartReplacements()
     {
-        return array(
+        return [
             '%product_value_conditions%' => implode(' OR ', $this->getProductValueConditions()),
-            '%product_value_joins%'      => implode(' ', $this->getProductValueJoins())
-        );
+            '%product_value_joins%'      => implode(' ', $this->getProductValueJoins()),
+            '%extra_joins%'              => implode(' ', $this->getExtraJoins()),
```
Comment (jjanvier) [inline src/Pim/Bundle/CatalogBundle/Doctrine/ORM/CompletenessGenerator.php]: I don't like the naming

File: src/Pim/Bundle/CatalogBundle/Doctrine/ORM/CompletenessGenerator.php [hunk 4]
```diff
@@ -346,10 +345,12 @@ protected function getMainSqlPart()
...
+            '%extra_conditions%'         => implode(' ', $this->getExtraConditions()),
```
Comment (jjanvier) [inline src/Pim/Bundle/CatalogBundle/Doctrine/ORM/CompletenessGenerator.php]: I don't like the naming

File: src/Pim/Bundle/CatalogBundle/Doctrine/ORM/CompletenessGenerator.php [hunk 5]
```diff
@@ -650,4 +666,29 @@ public function scheduleForChannelAndLocale(ChannelInterface $channel, LocaleInt
 
         $stmt->execute();
     }
+
+    /**
+     * @return string[]
+     */
+    protected function getExtraJoins()
```
Comment (jjanvier) [inline src/Pim/Bundle/CatalogBundle/Doctrine/ORM/CompletenessGenerator.php]: this is not an "extra" 
this is a join on prices

why do you need that ? to be overriden ?

File: src/Pim/Bundle/CatalogBundle/Doctrine/ORM/CompletenessGenerator.php [hunk 6]
```diff
@@ -650,4 +666,29 @@ public function scheduleForChannelAndLocale(ChannelInterface $channel, LocaleInt
...
+
+        $pricesJoin = sprintf($pricesJoin, static::COMPLETE_PRICES_TABLE);
+
+        return [$pricesJoin];
+    }
+
+    /**
+     * @return string[]
+     */
+    protected function getExtraConditions()
```
Comment (jjanvier) [inline src/Pim/Bundle/CatalogBundle/Doctrine/ORM/CompletenessGenerator.php]: same here

File: src/Pim/Bundle/CatalogBundle/Doctrine/ORM/CompletenessGenerator.php (hunk 5 above)
Comment (jmleroux) [inline src/Pim/Bundle/CatalogBundle/Doctrine/ORM/CompletenessGenerator.php]: Exactly

File: src/Pim/Bundle/CatalogBundle/Doctrine/ORM/CompletenessGenerator.php (hunk 6 above)
Comment (jmleroux) [inline src/Pim/Bundle/CatalogBundle/Doctrine/ORM/CompletenessGenerator.php]: Same answer ;)

**Pull request merged**